src/training/day5/Main.java

```diff
@@ -1,44 +1,179 @@
+                for (Person person : people) {
+                        if (person instanceof Student) {
+                                Student student = (Student) person;
+                                if (student.hasGradeForSubject(subject)) {
+                                        double grade = student.getHighestGrade();
+                                        if (grade > highestGrade) {
+                                                studentWithHighestGrade = student;
+                                                highestGrade = grade;
+                                        }
+                                }
+                        }
+                }
+                return studentWithHighestGrade;
+        }
+    public static Student findStudentWithLowestGrade(List<Person> people, String subject) {
+                Student studentWithLowestGrade = null;
+                double lowestGrade = Double.MAX_VALUE;
+                for (Person person : people) {
+                        if (person instanceof Student) {
+                                Student student = (Student) person;
+                                if (student.hasGradeForSubject(subject)) {
+                                        double grade = student.getLowestGrade();
+                                        if (grade < lowestGrade) {
+                                                studentWithLowestGrade = student;
+                                                lowestGrade = grade;
+                                        }
+                                }
+                        }
+                }
+                return studentWithLowestGrade;
+        }
+    public static double calculateAverageSalary(List<Person> people) {
+                double totalSalary = 0.0;
+                int teacherCount = 0;
+                for (Person person : people) {
+                        if (person instanceof Teacher) {
+                                Teacher teacher = (Teacher) person;
+                                totalSalary += teacher.getSalary();
+                                teacherCount++;
+                        }
+                }
+                if (teacherCount > 0) {
+                        return totalSalary / teacherCount;
+                }
+                return 0.0;
+
+        }
+    public static double calculateAverageClassGrade(List<Person> people) {
+                double totalGrade = 0.0;
+                int studentCount = 0;
+                for (Person person : people) {
+                        if (person instanceof Student) {
+                                Student student = (Student) person;
+                                totalGrade += student.getAverageGrade();
+                                studentCount++;
+                        }
+                }
+                if (studentCount > 0) {
+                        return Math.round( (totalGrade / studentCount) * 100.0) / 100.0;
```

Comment:
Format as: return Math.round((totalGrade / studentCount) * 100.0) / 100.0;
But it is very nice.